bftengine/src/bftengine/ReplicaFactory.cpp

```diff
@@ -0,0 +1,290 @@
+    bft::communication::ICommunication *communication,
+    MetadataStorage *metadataStorage,
+    std::shared_ptr<concord::performance::PerformanceManager> pm,
+    const shared_ptr<concord::secretsmanager::ISecretsManagerImpl> &sm) {
+  shared_ptr<PersistentStorage> persistentStoragePtr;
+  if (replicaConfig.debugPersistentStorageEnabled)
+    if (metadataStorage == nullptr)
+      persistentStoragePtr = std::make_unique<impl::DebugPersistentStorage>(replicaConfig.fVal, replicaConfig.cVal);
+
+  // Testing/real metadataStorage passed.
+  uint16_t numOfObjects = 0;
+  bool isNewStorage = true;
+  if (metadataStorage != nullptr) {
+    persistentStoragePtr.reset(new impl::PersistentStorageImp(
+        replicaConfig.numReplicas,
+        replicaConfig.fVal,
+        replicaConfig.cVal,
+        replicaConfig.numReplicas + replicaConfig.numRoReplicas + replicaConfig.numOfClientProxies +
+            replicaConfig.numOfExternalClients + replicaConfig.numOfClientServices + replicaConfig.numReplicas,
+        replicaConfig.clientBatchingMaxMsgsNbr));
+    unique_ptr<MetadataStorage> metadataStoragePtr(metadataStorage);
+    auto objectDescriptors = static_cast<PersistentStorageImp *>(persistentStoragePtr.get())
+                                 ->getDefaultMetadataObjectDescriptors(numOfObjects);
+    isNewStorage = metadataStoragePtr->initMaxSizeOfObjects(objectDescriptors, numOfObjects);
+    // Check if we need to remove the metadata or start a new epoch
+    bool erasedMetaData = false;
+    uint32_t actualObjectSize = 0;
+    metadataStoragePtr->read(ConstMetadataParameterIds::ERASE_METADATA_ON_STARTUP,
+                             sizeof(erasedMetaData),
+                             (char *)&erasedMetaData,
+                             actualObjectSize);
+    bool startNewEpoch = false;
+    metadataStoragePtr->read(
+        ConstMetadataParameterIds::START_NEW_EPOCH, sizeof(startNewEpoch), (char *)&startNewEpoch, actualObjectSize);
+    if (startNewEpoch) {
+      bftEngine::EpochManager::instance().startNewEpoch();
+      LOG_INFO(GL, "We should start a new epoch");
+    }
+    LOG_INFO(GL, KVLOG(erasedMetaData));
+    if (erasedMetaData) {
+      // Here when metadata is erased, we need to update DBCheckpointManager.
+      DbCheckpointManager::instance().setIsMetadataErased(true);
+      metadataStoragePtr->eraseData();
+      isNewStorage = metadataStoragePtr->initMaxSizeOfObjects(objectDescriptors, numOfObjects);
+      auto secFileDir = ReplicaConfig::instance().getkeyViewFilePath();
+      LOG_INFO(GL, "removing " << secFileDir << " files if exist");
+      try {
+        for (auto &it : fs::directory_iterator{secFileDir}) {
+          if (it.path().string().find(secFilePrefix) != std::string::npos) {
+            fs::remove(it.path());
+          }
+        }
+      } catch (std::exception &e) {
+        LOG_FATAL(GL, "unable to remove the secret file, as we erased the metadata we won't be able to restart");
```

Comment:
I never trust a logging system to terminate for me.
LOG_FATAL wont terminate.
We need something explicit. Since we catch exception here, I suggest to add a line after:
ConcordAssert(false)